toolchain/extensions.bzl

```
@@ -0,0 +1,41 @@
+"Module extensions for use with bzlmod"
+
+load("@com_grail_bazel_toolchain//toolchain:rules.bzl", "llvm_toolchain")
+load(
+    "@com_grail_bazel_toolchain//toolchain/internal:repo.bzl",
+    _llvm_config_attrs = "llvm_config_attrs",
+    _llvm_repo_attrs = "llvm_repo_attrs",
+)
+
+def _llvm_impl_(module_ctx):
+    for mod in module_ctx.modules:
+        for toolchain_attr in mod.tags.toolchain:
```

I would suggest going with something like this:

```suggestion
        for toolchain_attr in mod.tags.toolchain:
            if not mod.is_root:
                fail("Only the root module can use the 'llvm' extension")
```

This avoid having to deal with the difficult questions around reconciling toolchain requirements between different modules until we have had some time to test drive this.

---

@fmeum there is actually a real issue with this; if a transitive module requires access to the llvm_toolchain_llvm repo it can't `use_repo` on the extension as it hits this fail and so can't make the toolchain that was created by the root module visible into the transitive module where it needs to be used.

---

@jsharpe Since this `fail` only applies if there is a `toolchain` tag, transitive modules should still be able to `use_extension` and `use_repo` this if they don't specify any tags. Are you seeing different behavior?

However, transitive modules of course can't know which name the root module chose for the toolchain. I would thus recommend to make `name` an optional attribute and have it default to something that transitive modules can hardcode (e.g. `llvm`). Ideally, the module would then *always* register that toolchain with some reasonable default even if the user doesn't, so that transitive modules can rely on it being present.

An alternative could be to expose a repo under a fixed name that contains a Starlark variable pointing to the first toolchain repo defined by the user. See https://github.com/bazelbuild/rules_go/blob/f03a723c49d5ec8afa0f9c74c4c3a705b37e8f40/go/private/extensions.bzl#L4-L21 for an example.

All of this depends on how you envision transitive modules using the toolchain: Should a transitive module be able to use the toolchain without the root module's cooperation? 

CC @siddharthab